bugtool/cmd/root.go

```diff
@@ -0,0 +1,297 @@
+)
+
+func init() {
+	RootCmd.Flags().BoolVar(&serve, "serve", false, "Serve captured files over HTTP")
+	RootCmd.Flags().IntVarP(&port, "port", "p",
+		4444, "Port to use when serving files, default is 4444")
+	RootCmd.Flags().StringVarP(&dumpPath, "tmp", "t", "/tmp/", "Path to store extraced files")
+}
+
+func runTool() {
+	var dbgDir string
+	// Prevent collision with other directories
+	d, err := ioutil.TempDir(dumpPath, "bugtool-")
+	if err != nil {
+		log.WithError(err).Warn("failed to create debug directory")
+	}
+	dbgDir = d
+
+	// TODO: review all error messages.
+	copyStateDir(dbgDir)
+	copySystemInfo(dbgDir)
+	// TODO: Copy relevent log files
+	// TODO: retrieve k8s log files
+
+	/*
+		Docker logs
+		access logs
+	*/
+
+	copyResolveConfig(dbgDir)
+	copyKernelConfig(dbgDir)
+	createArchive(dbgDir, dumpPath)
+
+	if serve {
+		serveStaticFiles(dbgDir)
+		return
+	}
+
+	if err := os.RemoveAll(dbgDir); err != nil {
+		log.WithError(err).Warn("failed to cleanup temporary files")
+	}
+}
+
+func copyStateDir(dbgDir string) {
+	src := filepath.Join(defaults.RuntimePath, defaults.StateDir)
+	dst := filepath.Join(dbgDir, defaults.StateDir)
+	if err := copyDir(src, dst); err != nil {
+		log.WithError(err).Warn("could not copy state directory")
+	}
+}
+
+func serveStaticFiles(debugDirectory string) {
+	fs := http.FileServer(http.Dir(debugDirectory))
+	http.Handle("/", fs)
+
+	log.Printf("Serving files at http://localhost:%d", port)
+	http.ListenAndServe(fmt.Sprintf(":%d", port), nil)
+}
+
+func copyFile(dir, from, prefix, suffix string) error {
+	// Read the file we want to copy
+	srcFile, err := os.OpenFile(from, os.O_RDONLY, 0600)
+	defer srcFile.Close()
+	if err != nil {
+		return err
+	}
+	// Create the new file and copy contents of original
+	dstFile := tempFile(dir, prefix)
+	if _, err := io.Copy(dstFile, srcFile); err != nil {
+		return err
+	}
+	dstFile.Close()
+	// Add the suffix so desktop programs identify the file
+	renameFileWithSuffix(dstFile.Name(), suffix)
+	return nil
+}
+
+func copyResolveConfig(dbgDir string) {
+	confDir := createDir(dbgDir, "conf")
+	src, _ := os.OpenFile("/etc/resolv.conf", os.O_RDONLY, 0600)
+	dst, _ := os.Create(filepath.Join(confDir, "resolv.conf"))
+	if _, err := io.Copy(dst, src); err != nil {
+		log.WithError(err).Warn("could not copy resolve config")
+	}
+	src.Close()
+	dst.Close()
+}
+
+func createDir(dbgDir string, newDir string) string {
+	confDir := filepath.Join(dbgDir, newDir)
+	if err := os.Mkdir(confDir, defaults.RuntimePathRights); err != nil {
+		log.WithError(err).Warn("failed to create %s info directory", newDir)
+		return dbgDir
+	}
+	return confDir
+}
+
+func copyKernelConfig(debugDirectory string) {
+	// TODO: add this
+}
+
+func createArchive(dbgDir, tmpDirPrefix string) {
+	progress := func(a string) {
+		fmt.Println(a)
+	}
+	// We are using ZIP format because it's widely supported.
+	output := filepath.Join(tmpDirPrefix, dbgDir+".zip")
+	zip.ArchiveFile(dbgDir, output, progress)
```

Comment:
I'm undecided here. On one side I like not relying on additional external tools. OTOH, this adds more external dependencies to our already large list. Maybe just invoke `tar` for this?

Reply:
I was using `tar` earlier but had some issue with directories where it would just create empty files. Can look into it again and revert.